Source/render/NodeRenderer.swift

```diff
@@ -69,22 +64,56 @@ class NodeRenderer {
         fatalError("Unsupported")
     }
 
-    final public func render(force: Bool, opacity: Double) {
-        ctx.cgContext!.saveGState()
+    final public func render(in context: CGContext, force: Bool, opacity: Double, useAlphaOnly: Bool = false) {
+        context.saveGState()
         defer {
-            ctx.cgContext!.restoreGState()
+            context.restoreGState()
         }
-
         guard let node = node() else {
             return
         }
+        let newOpacity = node.opacity * opacity
+
+        context.concatenate(node.place.toCG())
+        applyClip(in: context)
+
+        // no effects, just draw as usual
+        guard let effect = node.effect else {
+            directRender(in: context, force: force, opacity: newOpacity, useAlphaOnly: useAlphaOnly)
+            return
+        }
 
-        ctx.cgContext!.concatenate(node.place.toCG())
-        applyClip()
-        directRender(force: force, opacity: node.opacity * opacity)
+        var effects = [Effect]()
+        var next: Effect? = effect
+        while next != nil {
+            effects.append(next!)
+            next = next?.input
+        }
+
+        let offset = effects.first { $0 is OffsetEffect } as? OffsetEffect
+        let otherEffects = effects.filter { !($0 is OffsetEffect) }
+        let useAlphaOnly = otherEffects.contains { effect -> Bool in
+            effect is AlphaEffect
+        }
+
+        // move to offset
+        let move = Transform(m11: 1, m12: 0, m21: 0, m22: 1, dx: offset?.dx ?? 0, dy: offset?.dy ?? 0)
```

Comment:
There is a better implementation:
```swift
if let ofst = offset {
    context.concatenate(CGAffineTransform(translationX: ofst.dx, y: ofst.dy)))
}
```
It much faster, because:
* you make 4 checks and still multiply matrix when there is no offset
* you create `Transform` while use `CGAffineTransform`
* don't use common transform constructor when you need translate. Modern matrix implementation works faster when you say exactly which transformation is used